ang/civiawards/directives/award.directive.js

```diff
@@ -0,0 +1,231 @@
+(function (angular, $, _) {
+  var module = angular.module('civiawards');
+
+  module.directive('civiaward', function () {
+    return {
+      scope: {
+        awardId: '='
+      },
+      controller: 'CiviAwardCreateEditAwardController',
+      templateUrl: '~/civiawards/directives/award.directive.html',
+      restrict: 'E'
+    };
+  });
+
+  module.controller('CiviAwardCreateEditAwardController', function ($scope, $window, crmApi, getSelect2Value, CaseStatus, AwardType, CaseTypeCategory) {
+    var ts = CRM.ts('civicase');
+
+    $scope.ts = ts;
+    $scope.isNameDisabled = true;
+    $scope.submitInProgress = false;
+    $scope.awardTypes = [];
+    $scope.awardDetailsID = null;
+    $scope.awardStages = CaseStatus.getAll();
+    $scope.tabs = [
+      { name: 'stages', label: ts('Award Stages') },
+      { name: 'additionaltab', label: ts('Additional Tab') }
+    ];
+    $scope.basicDetails = {
+      title: '',
+      name: '',
+      description: '',
+      awardType: null,
+      isEnabled: true,
+      startDate: null,
+      endDate: null,
+      awardManagers: [],
+      selectedAwardStages: {}
+    };
+
+    $scope.createNewAwardStage = createNewAwardStage;
+    $scope.saveAward = saveAward;
+    $scope.selectTab = selectTab;
+
+    (function init () {
+      $scope.activeTab = $scope.tabs[0].name;
+
+      mapAwardTypes();
+
+      $scope.$watch('basicDetails.title', titleWatcher);
+
+      if ($scope.awardId) {
+        fetchAwardInformation()
+          .then(function (result) {
+            setBasicDetails(result.caseType);
+            setAdditionalInformation(result.additionalDetails);
+          });
+      } else {
+        enableAllAwardStage();
+      }
+    }());
+
+    /**
+     * Watcher for Title form field
+     */
+    function titleWatcher () {
+      if (!$scope.awardId && $scope.basic_details_form.awardName.$pristine) {
+        $scope.basicDetails.name = $scope.basicDetails.title.replace(/ /g, '_').replace(/[^a-zA-Z0-9_]/g, '').toLowerCase();
+      }
+    }
+
+    /**
+     * Set basic details
+     *
+     * @param {object} caseType case type
+     */
+    function setBasicDetails (caseType) {
+      $scope.basicDetails.title = caseType.title;
+      $scope.basicDetails.name = caseType.name;
+      $scope.basicDetails.description = caseType.description;
+      $scope.basicDetails.isEnabled = caseType.is_active === '1';
+
+      _.each(caseType.definition.statuses, function (stageName) {
+        var awardStage = _.find($scope.awardStages, function (stage) {
+          return stage.name === stageName;
+        });
+        $scope.basicDetails.selectedAwardStages[awardStage.value] = true;
```

Comment:
> what happens if selectedAwardStages already had values selected. Is it ok to not reset this var?

not sure, what you mean. This is happening on init, so how come it will already have some values.

Also refactored the function like this

```javascript
setAwardStages(caseType.definition.statuses);

/**
 * get Selected Award stages
 *
 * @param {Array} awardStages award stages fetched from api
 */
function setAwardStages (awardStages) {
  _.each(awardStages, function (stageName) {
    var awardStage = _.find($scope.awardStages, function (stage) {
      return stage.name === stageName;
    });
    $scope.basicDetails.selectedAwardStages[awardStage.value] = true;
  });
}
```

because if I set the value directly on `$scope.basicDetails.selectedAwardStages` instead of  returning another object, its more efficient.

Reply:
@deb1990 the problem with these type of functions is that they are modifying a variable and it's not clear that this is being done.

Performance-wise I don't think an `each` has an advantage over `reduce`. It does about the same process:
```js
function getSelectedAwards (awardStages) {
  return _.reduce(awardStages, function (selectedAwards, stageName) {
    var awardStage = _.find($scope.awardStages, function (stage) {
      return stage.name === stageName;
    });

    selectedAwards[awardStage.value] = true;

    return selectedAwards;
  }, {});
}
```

The only difference is that the final result is passed down. But I wouldn't be worried about performance here unless it's something considerable noticeable.

There's a bonus to getting the awards instead of setting them: the value gets reset each time. What happens if we call `setAwardStages` multiple times with different parameters? It would preserve the previous value. I know this is not part of the current workflow, but that's a flag.

Reply:
Okay makes sense. I have changed the code, but have not used `reduce`, used `each` instead.

Reply:
@jamienovick 

Resolved comments in the spec say that we do not need a page for Case type ie we can build on top of the current screen (which isn't the best UX).

Should create this as in the wireframe ie a bit like events?